File: ios/MarkerImageLoader.swift
import Foundation
import UIKit

// Loads a marker icon from the URI the JS layer resolved (via Image.resolveAssetSource):
//   - http(s):// — fetched off the main thread (Metro dev-server assets arrive this way),
//   - data:image;base64 — decoded inline,
//   - file:// — read from disk,
//   - otherwise treated as a bundled image name (release builds of require()d assets).
// Decoded images are cached by URI; the completion always runs on the main queue (where MapKit
// objects must be touched).
internal enum MarkerImageLoader {
  private static let cache = NSCache<NSString, UIImage>()

  static func load(_ source: String, completion: @escaping (UIImage?) -> Void) {
    if let cached = cache.object(forKey: source as NSString) {
      completion(cached)
      return
    }
    DispatchQueue.global(qos: .userInitiated).async {
      let image = decode(source)
      DispatchQueue.main.async {
        if let image = image {
          cache.setObject(image, forKey: source as NSString)
        }
        completion(image)
      }
    }
  }

  private static func decode(_ source: String) -> UIImage? {
    if source.hasPrefix("http://") || source.hasPrefix("https://") {
      guard let url = URL(string: source), let data = try? Data(contentsOf: url) else {
        return nil
      }
      return UIImage(data: data)
    }
    if source.hasPrefix("data:image") {
      guard let comma = source.firstIndex(of: ","),
        let data = Data(base64Encoded: String(source[source.index(after: comma)...]))
      else {
        return nil
      }
      return UIImage(data: data)
    }
    if source.hasPrefix("file://") {
      return UIImage(contentsOfFile: String(source.dropFirst("file://".count)))
    }
    return UIImage(named: source)
  }
}

File: android/src/main/java/expo/modules/yandexmapkit/MarkerImageLoader.kt
package expo.modules.yandexmapkit

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.os.Handler
import android.os.Looper
import android.util.Base64
import android.util.Log
import java.net.URL
import java.util.concurrent.ConcurrentHashMap

// Loads a marker icon from the URI the JS layer resolved (via Image.resolveAssetSource):
//   - http(s):// — fetched off the main thread (Metro dev-server assets arrive this way),
//   - data:image;base64 — decoded inline,
//   - file:// — read from disk,
//   - otherwise treated as a drawable resource name (release builds of bundled require()d assets).
// Decoded bitmaps are cached by URI so re-rendering a marker with the same icon costs nothing.
// The callback always runs on the main thread (where MapKit objects must be touched).
internal object MarkerImageLoader {
  private const val TAG = "ExpoYandexMapKit"
  private val cache = ConcurrentHashMap<String, Bitmap>()
  private val mainHandler = Handler(Looper.getMainLooper())

  fun load(context: Context, source: String, callback: (Bitmap?) -> Unit) {
    cache[source]?.let {
      callback(it)
      return
    }
    val appContext = context.applicationContext
    Thread {
      val bitmap = try {
        decode(appContext, source)
      } catch (e: Exception) {
        Log.w(TAG, "Failed to load marker image \"$source\"", e)
        null
      }
      mainHandler.post {
        if (bitmap != null) {
          cache[source] = bitmap
        }
        callback(bitmap)
      }
    }.start()
  }

  private fun decode(context: Context, source: String): Bitmap? {
    return when {
      source.startsWith("http://") || source.startsWith("https://") ->
        URL(source).openStream().use { BitmapFactory.decodeStream(it) }

      source.startsWith("data:image") -> {
        val base64 = source.substringAfter(',', "")
        val bytes = Base64.decode(base64, Base64.DEFAULT)
        BitmapFactory.decodeByteArray(bytes, 0, bytes.size)
      }

      source.startsWith("file://") ->
        BitmapFactory.decodeFile(source.removePrefix("file://"))

      else -> {
        val resId = context.resources.getIdentifier(source, "drawable", context.packageName)
        if (resId == 0) null else BitmapFactory.decodeResource(context.resources, resId)
      }
    }
  }
}
